src/app/user-form/user-form.component.ts

```diff
@@ -0,0 +1,136 @@
+import { Component, EventEmitter, Input, OnChanges, OnInit, Output, SimpleChanges } from '@angular/core';
+import { AbstractControl, FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { User } from '../user';
+import { UserService } from '../user.service';
+import { EmailValidator } from '../email.validator';
+
+@Component({
+	selector: 'app-user-form',
+	templateUrl: './user-form.component.html',
+	styleUrls: ['./user-form.component.css']
+})
+export class UserFormComponent implements OnInit, OnChanges {
+
+	@Input()
+	public formType: "Add" | "Edit" = "Add";
+
+	@Input()
+	public userToEdit: User | undefined = undefined;
+
+	@Output()
+	public submitFormEvent: EventEmitter<{ user: User, formType: "Add" | "Edit" }> = new EventEmitter<{ user: User, formType: "Add" | "Edit" }>();
+
+	@Output()
+	public closeFormEvent: EventEmitter<boolean> = new EventEmitter<boolean>();
+
+	public isFormOpen: boolean = true;
+
+	public userForm: FormGroup = this.formBuilder.group({
+		firstName: ["", [Validators.required, Validators.minLength(3)]],
+		lastName: ["", [Validators.required, Validators.minLength(3)]],
+		email: ["", Validators.compose([
+			Validators.required,
+			Validators.email,
+			EmailValidator.validateEmail(this.userService, { duplicateEmail: true })
+		])]
+	})
+
+	constructor(
+		private formBuilder: FormBuilder,
+		private userService: UserService
+	) {
+		this.userForm = this.constructForm();
+	}
+
+	ngOnChanges(changes: SimpleChanges): void {
+		if (changes['formType']) {
+			this.formType = changes['formType'].currentValue;
+		}
+		this.userToEdit = changes['userToEdit'].currentValue;
+		this.userForm = this.constructForm();
+
+		if (this.formType == "Edit") {
+			if (this.userToEdit) {
+				this.populateForm(this.userToEdit);
+			}
+		}
+	}
+
+	ngOnInit(): void {
+		if (this.formType === "Edit") {
+			if (this.userToEdit) {
+				this.populateForm(this.userToEdit);
+			}
+		}
+	}
+
+	public get firstNameFormControl(): AbstractControl | null {
+		return this.userForm.get('firstName');
+	}
+
+	public get lastNameFormControl(): AbstractControl | null {
+		return this.userForm.get("lastName");
+	}
+
+	public get emailFormControl(): AbstractControl | null {
+		return this.userForm.get("email");
+	}
+
+	private constructForm(): FormGroup {
+		return this.formBuilder.group({
+			firstName: ["", [Validators.required, Validators.minLength(3)]],
+			lastName: ["", [Validators.required, Validators.minLength(3)]],
+			email: ["", Validators.compose([
+				Validators.required,
+				Validators.email,
+				EmailValidator.validateEmail(this.userService, { duplicateEmail: true })
+			])]
+		});
+	}
+
+	private populateForm(userToEdit: User): void {
```

Comment:
```suggestion
	private populateForm(): void {
```
you have userToEdit globally